src/query/pipeline/core/src/finished_chain.rs

```diff
@@ -0,0 +1,388 @@
+                    location,
+                    is_always: false,
+                    successfully: false,
+                    elapsed: instant.elapsed(),
+                });
+
+                info.res = Err(cause.clone());
+                Self::apply_always(info, states, always_callbacks);
+
+                return Err(cause);
+            }
+
+            states.push(ApplyState {
+                location,
+                is_always: false,
+                successfully: true,
+                elapsed: instant.elapsed(),
+            });
+        }
+
+        Self::apply_always(info, states, always_callbacks);
+        Ok(())
+    }
+
+    fn apply_always(
+        info: ExecutionInfo,
+        mut states: Vec<ApplyState>,
+        always_callbacks: Vec<(&'static Location<'static>, Box<dyn Callback>)>,
+    ) {
+        for (location, always_callback) in always_callbacks {
+            let instant = Instant::now();
+            states.push(match always_callback.apply(&info) {
+                Ok(_) => ApplyState {
+                    location,
+                    is_always: true,
+                    successfully: true,
+                    elapsed: instant.elapsed(),
+                },
+                Err(_cause) => ApplyState {
+                    location,
+                    is_always: true,
+                    successfully: false,
+                    elapsed: instant.elapsed(),
+                },
+            });
+        }
+
+        Self::log_states(&states);
+    }
+
+    pub fn extend(&mut self, other: FinishedCallbackChain) {
+        self.chain.extend(other.chain)
+    }
+
+    fn log_states(apply_states: &[ApplyState]) {
+        let mut message = String::new();
+        writeln!(&mut message, "Executor apply finished callback state:").unwrap();
+        for apply_state in apply_states {
+            let execute_state = match apply_state.successfully {
+                true => "\u{2705}",
+                false => "\u{274C}",
+            };
+
+            let always_state = match apply_state.is_always {
+                true => "(always) ",
+                false => "",
+            };
+
+            writeln!(
+                &mut message,
+                "├──{}:{:?} - {}{}:{}:{}",
+                execute_state,
+                apply_state.elapsed,
+                always_state,
+                apply_state.location.file(),
+                apply_state.location.line(),
+                apply_state.location.column()
+            )
+            .unwrap();
+        }
+
+        info!("{}", message);
+    }
+}
+
+impl<T: FnOnce(&ExecutionInfo) -> Result<()> + Send + Sync + 'static> Callback for T {
+    fn apply(self: Box<Self>, info: &ExecutionInfo) -> Result<()> {
+        match catch_unwind(move || self(info)) {
+            Ok(Ok(_)) => Ok(()),
+            Err(cause) => Err(cause),
+            Ok(Err(cause)) => Err(cause),
+        }
+    }
+}
+
+pub struct AlwaysCallback<T: Callback> {
```

Comment:
except for  `always`, do we need to add a flag to distinguish whether a callback failure will cause the statement to fail? Or should we just explain the current behavior through comments?”

Reply:
`always` is special, by default, callback failure should always cause the statement to fail.

Reply:
so `always` include 3 meanings

1.  itself always gets run, not affected by others 
2. if it fail, not affect others
3. if it fail, not affect a result of the stament.